### PR TITLE
Modifications to support plugin refactoring

### PR DESCRIPTION
**Pull request by charles-cowart**

### CODE REVIEW

File: sequence_processing_pipeline/Job.py [hunk 1]
```diff
@@ -150,6 +150,10 @@ def _system_call(self, cmd, allow_return_codes=[], callback=None):
         :return: a dictionary containing stdout, stderr, and return_code as
                  key/value pairs.
         """
+        # when system-wide binaries such as 'sbatch' and 'sacct' are not
+        # present, assume executable w/no path is present in current
+        # directory.
+        cmd = 'export PATH="$PATH:.";' + cmd
```
Comment (wasade) [inline sequence_processing_pipeline/Job.py:156]: Why not:

```suggestion
        cmd = './' + cmd
```

**Comment by charles-cowart** [comment]
@antgonza Ready for review

File: sequence_processing_pipeline/Pipeline.py [hunk 2]
```diff
@@ -52,6 +53,11 @@ def __init__(self, configuration_file_path, run_id, sample_sheet_path,
             raise PipelineError("sample_sheet_path or mapping_file_path "
                                 "must be defined, but not both.")
 
+        if type not in ['amplicon', 'metagenomic', 'metatranscriptomic']:
```
Comment (wasade) [inline sequence_processing_pipeline/Pipeline.py:56]: Could these be centralized as static strings?

`type` is a reserved word in Python. Could a variation be used, such as `type_` or `run_type`?

File: sequence_processing_pipeline/Pipeline.py [hunk 3]
```diff
@@ -338,15 +344,28 @@ def get_project_info(self):
                                   self.mapping_file.groupby('project_name')}
 
             for project in sample_project_map:
-                qiita_id = project.split('_')[-1]
+                matches = search(r'(.+)_(\d+)$', str(project))
```
Comment (wasade) [inline sequence_processing_pipeline/Pipeline.py:347]:
```suggestion
                matches = search(r'^(.+)_(\d+)$', str(project))
```

File: sequence_processing_pipeline/Pipeline.py [hunk 4]
```diff
@@ -338,15 +344,28 @@ def get_project_info(self):
...
+                qiita_id = matches[2]
```
Comment (wasade) [inline sequence_processing_pipeline/Pipeline.py:348]: I believe this will throw an `IndexError` if `search` does not find matches as would happen on a malformed `project`

Comment (wasade) [reply sequence_processing_pipeline/Pipeline.py:348]: It may be valuable to decompose the `search` operation into a discrete function that handles edge cases. It then also makes it very easy to enumerate tests of the regex, with a pattern like

```python
def test_extract_sample_project(self):
    tests = [('foo_123', ('foo', '123')), ('bar.baz_123', ('bar.baz', '123')), ('bad', None), ...etc...]
    for test, exp in tests:
        if exp is None:
            with self.assertRaises(InvalidProject):
                extract_sample_project(test)
        else:
            obs = extract_sample_project(test)
            self.assertEqual(obs, exp)
```

File: sequence_processing_pipeline/Pipeline.py [hunk 5]
```diff
@@ -338,15 +344,28 @@ def get_project_info(self):
...
             bioinformatics = self.sample_sheet.Bioinformatics
             for result in bioinformatics.to_dict('records'):
-                results.append({'project_name': result['Sample_Project'],
-                                'qiita_id': result['QiitaID']})
+                qiita_id = result['QiitaID']
+                project = result['Sample_Project']
+
+                if short_names:
+                    matches = search(r'(.+)_(\d+)$', str(project))
+                    project = matches[1]
```
Comment (wasade) [inline sequence_processing_pipeline/Pipeline.py:364]: same comments here

File: sequence_processing_pipeline/tests/test_Pipeline.py [hunk 6]
```diff
@@ -171,7 +170,7 @@ def test_creation(self):
                      self.good_run_id,
                      self.bad_assay_type_path, None,
                      self.output_file_path,
-                     self.qiita_id, None)
+                     self.qiita_id, 'metagenomic', None)
```
Comment (wasade) [inline sequence_processing_pipeline/tests/test_Pipeline.py:173]: I'll just note here, not on the below, but it would be ideal for this to be the static variable suggested in an earlier comment

**Review by wasade — CHANGES_REQUESTED** [review]

File: sequence_processing_pipeline/Job.py (hunk 1 above)
Comment (charles-cowart) [reply sequence_processing_pipeline/Job.py:156]: When the code is running in production or qiita-rc, the executable will be loaded in via linux module and be available on the path. When running in unittests, a mock executable will be created in the current directory. I believe if we simply prepend './' to the cmd it would not locate the binaries in production.

File: sequence_processing_pipeline/Pipeline.py (hunk 2 above)
Comment (charles-cowart) [reply sequence_processing_pipeline/Pipeline.py:56]: Good catch, thanks!

File: sequence_processing_pipeline/Job.py (hunk 1 above)
Comment (charles-cowart) [reply sequence_processing_pipeline/Job.py:156]: I considered it. This was the quickest solution and it stuck. I'm happy to revisit it however.

Comment (wasade) [reply sequence_processing_pipeline/Job.py:156]: Can't something equivalent to `export PATH=$(pwd):$PATH` be added to github actions and that solves the issue?

Comment (charles-cowart) [reply sequence_processing_pipeline/Job.py]: Probably.

**Comment by charles-cowart** [comment]
@wasade Thanks for the review! Ready for re-review.

File: sequence_processing_pipeline/Job.py [hunk 7]
```diff
@@ -150,6 +150,9 @@ def _system_call(self, cmd, allow_return_codes=[], callback=None):
         :return: a dictionary containing stdout, stderr, and return_code as
                  key/value pairs.
         """
+        # when system-wide binaries such as 'sbatch' and 'sacct' are not
+        # present, assume executable w/no path is present in current
+        # directory.
```
Comment (wasade) [inline sequence_processing_pipeline/Job.py:155]: Is this comment still needed? It doesn't look like a change to the test runner was necessary?

File: sequence_processing_pipeline/Pipeline.py [hunk 8]
```diff
@@ -31,8 +32,15 @@ class Pipeline:
                     None, 32.5, -117.25, 'control blank', 'metagenome', 256318,
                     None, 'adaptation', 'TRUE', 'UCSD', 'FALSE']
 
+    METAGENOMIC_PTYPE = 'metagenomic'
+    METATRANSCRIPTOMIC_PTYPE = 'metatranscriptomic'
+    AMPLICON_PTYPE = 'amplicon'
+
+    pipeline_types = [METAGENOMIC_PTYPE, AMPLICON_PTYPE,
+                      METATRANSCRIPTOMIC_PTYPE]
```
Comment (wasade) [inline sequence_processing_pipeline/Pipeline.py:40]:
```suggestion
    pipeline_types = {METAGENOMIC_PTYPE, AMPLICON_PTYPE,
                      METATRANSCRIPTOMIC_PTYPE}
```

**Review by wasade — APPROVED** [review]
Just a minor question, thank you @charles-cowart!

**Review by antgonza — APPROVED** [review]
Looks good and agree with @wasade's comments; thank you both!

**Review by antgonza — APPROVED** [review]
Looks good and agree with @wasade's comments; thank you both!

File: sequence_processing_pipeline/Job.py (hunk 7 above)
Comment (charles-cowart) [reply sequence_processing_pipeline/Job.py:155]: I'm going to try adding it to the environment on the qp-klp side of things. If something's needed I'll introduce it, I will.
